### PR TITLE
cmd-ref: Document ls-url command

### LINKED ISSUE #3992: Document `dvc ls-url`
### Report

`dvc ls-url` is a new command being implemented in iterative/dvc#8299

### PR DESCRIPTION
**Pull request by rlamy**
Document `ls-url` command, which is being implemented in iterative/dvc#8299.

Fixes #3992 

In review app: https://dvc-org-dvc-ls-url-jhwjeyjd4mh.herokuapp.com/doc/command-reference/list-url

### CODE REVIEW

File: content/docs/command-reference/list-url.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+inside `data`.
+
+</details>
+
+<details>
+
+### local
+
+```dvc
+$ dvc get-url /local/path/to/data
```
Comment (dberenbaum) [inline content/docs/command-reference/list-url.md:92]:
```suggestion
$ dvc ls-url /local/path/to/data
```

**Review by dberenbaum — COMMENTED** [review]
Looks great! Just one minor typo.

**Review by shcheklein — APPROVED** [review]

**Comment by shcheklein** [comment]
need to update syntax highlighter I think ... @rogermparent @yathomasi could you pls help / guide where and how it can be done pls?

**Comment by rogermparent** [comment]
We've been working on the syntax highlighter, last I remember @yathomasi is on it right now and we're parked at a spot where we don't currently have a good way to grab/update the highlight definitions we share for dvc/cml/mlem since they're generated from the `sidebar.json` system and are difficult to share as a result. It's already pretty high priority, but we can bump it higher since it seems like the issue people are running into most.

File: content/docs/sidebar.json [hunk 2]
```diff
@@ -323,6 +323,10 @@
         "label": "install",
         "slug": "install"
       },
+      {
+        "label": "list-url",
+        "slug": "list-url"
+      },
       {
         "label": "list",
         "slug": "list"
```
Comment (jorgeorpinel) [inline content/docs/sidebar.json:332]: Please put it after `list` for consistency with existing sorting.

File: content/docs/command-reference/list-url.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
+# list-url
+
+> Aliased to `dvc ls-url`
+
+List contents from a supported URL (for example `s3://`, `ssh://`, and other
+protocols).
+
+> Useful to find data to `dvc get-url` or `dvc import-url`.
```
Comment (jorgeorpinel) [inline content/docs/command-reference/list-url.md:8]: I'd link back from those other refs to this one too.

Feel free to copy the notes from the ref pages of `get` and `import` (which mention `list`).

Comment (jorgeorpinel) [inline content/docs/command-reference/list-url.md:8]: Minor but ideally use proper admonitions for notes. I realize that this is copied from `list` (written before we had `<admon>`) and that it's @rlamy's first PR so here's what I mean:

```suggestion
<admon type="info">

Aliased to `dvc ls-url`

</admon>

List contents from a supported URL (for example `s3://`, `ssh://`, and other
protocols).

<admon type="tip">

Useful to find data to `dvc get-url` or `dvc import-url`.

</admon>
```

Related: https://github.com/iterative/dvc.org/issues/3382
More context: https://github.com/iterative/cml.dev/pull/210

File: content/docs/command-reference/list-url.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+## Synopsis
+
+```usage
+usage: dvc ls-url [-h] [-q | -v] [-R] url
+
+positional arguments:
+  url              See `dvc get-url` for full list of supported URLs.
```
Comment (jorgeorpinel) [inline content/docs/command-reference/list-url.md:16]:
```suggestion
usage: dvc list-url [-h] [-q | -v] [-R] url

positional arguments:
  url              See `dvc list-url` for full list of supported URLs.
```
☝🏼 Seems copy/pasted from `get-url` ref. Is this the actual output of `list-url -h`?

File: content/docs/command-reference/list-url.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+```
+
+## Description
+
+Lists files and directories from an external location. `dvc ls-url` is
+equivalent to a wrapper providing a uniform interface around various listing
+commands such as `aws s3 ls` or `ssh user@host ls -a`.
```
Comment (jorgeorpinel) [inline content/docs/command-reference/list-url.md:23]: Minor, not a strong opinion but:
```suggestion
Lists files and directories from an external location. `dvc ls-url` is
a wrapper providing a uniform interface around various equivalent listing
commands such as `aws s3 ls` or `ssh user@host ls -a`.
```

File: content/docs/command-reference/list-url.md [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+> `[azure]`, `[gs]`, `[oss]`, `[ssh]`. Alternatively, use `[all]` to include
+> them all. The command should look like this: `pip install "dvc[s3]"`. (This
+> example installs `boto3` library along with DVC to support S3 storage.)
+
+Only the root directory is listed by default, but the `-R` option can be used to
+list files recursively.
+
+## Options
+
+- `-R`, `--recursive` - recursively list files of all subdirectories.
```
Comment (jorgeorpinel) [inline content/docs/command-reference/list-url.md:47]: Minor: Maybe use the same wording as in https://dvc.org/doc/command-reference/list#options ? I.e.

```suggestion
- `-R`, `--recursive` - recursively prints contents of all subdirectories.
```

> Or perhaps update the `list` ref for language consistency.

File: content/docs/command-reference/list-url.md [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+- `-q`, `--quiet` - do not write anything to standard output. Exit with 0 if no
+  problems arise, otherwise a non-zero value.
+
+- `-v`, `--verbose` - displays detailed tracing information.
+
+## Examples
+
+<details>
+
+### Amazon S3
```
Comment (jorgeorpinel) [inline content/docs/command-reference/list-url.md:60]: @dberenbaum are you sure you don't want to keep the existing example format from all other refs? Keep in mind that H2s are listed on the right-side bar. I.e.

```suggestion
## Example: Amazon S3
```

File: content/docs/command-reference/list-url.md (hunk 1 above)
Comment (jorgeorpinel) [reply content/docs/command-reference/list-url.md]: I committed this suggestion ☝🏼 

> @rlamy please `git pull` on your end.

File: content/docs/sidebar.json (hunk 2 above)
Comment (jorgeorpinel) [reply content/docs/sidebar.json:332]: > put it after list for consistency

Never mind, I was confused about this one. Resolving comment...

**Comment by yathomasi** [comment]
For alias please update: https://github.com/iterative/dvc.org/blob/main/content/linked-terms.js
For the addition of new commands: 
	- we need to update respective command file [here](https://github.com/iterative/gatsby-theme-iterative/tree/main/packages/gatsby-theme-iterative/config/prismjs)
	- I did write a [rough script](https://github.com/iterative/gatsby-theme-iterative/pull/18/files#diff-6335516df4ca9ea338a7966639c72d56013240d6b92affc717f53e4e1d6ff338) to update the commands a while back expecting this same situation. Now, we can bump up to higher priority and I think we can update this into GHA to create PR with CML.


The one @rogermparent mentioned is for cross-websites command line linker. [draft PR](https://github.com/iterative/gatsby-theme-iterative/pull/70)
> We've been working on the syntax highlighter, last I remember @yathomasi is on it right now and we're parked at a spot where we don't currently have a good way to grab/update the highlight definitions we share for dvc/cml/mlem since they're generated from the `sidebar.json` system and are difficult to share as a result. It's already pretty high priority, but we can bump it higher since it seems like the issue people are running into most.

File: content/docs/command-reference/list-url.md (hunk 7 above)
Comment (dberenbaum) [reply content/docs/command-reference/list-url.md:60]: Sure, sounds good. Thanks @jorgeorpinel!

Comment (dberenbaum) [reply content/docs/command-reference/list-url.md:72]: @rlamy Do you mind making this change?

Comment (rlamy) [reply content/docs/command-reference/list-url.md:72]: Sure, I'll do it.

File: content/docs/command-reference/list-url.md (hunk 5 above)
Comment (rlamy) [reply content/docs/command-reference/list-url.md]: It doesn't actually wrap any of these commands, it just behaves similarly, hence the clunky "equivalent to a wrapper". I guess I should rephrase the whole sentence.

File: content/docs/command-reference/list-url.md (hunk 6 above)
Comment (rlamy) [reply content/docs/command-reference/list-url.md:59]: I want to emphasise that the command only returns files (not directories). I updated the `list` ref which wrongly implied that `dvc ls -R` returns directories.

**Comment by rlamy** [comment]
`dvc list-url` has been merged. Could someone do a final review please?

@yathomasi Should I send a PR simply adding `list-url` to [gatsby-theme-iterative](https://github.com/iterative/gatsby-theme-iterative)?

**Comment by rogermparent** [comment]
Thomas is on vacation, so I'll help you see this through. We're just adding `ls-url` and `list-url` to `dvc`, right? I opened up [this PR](https://github.com/iterative/gatsby-theme-iterative/pull/89) for that. Sorry for not helping out for so long, I actually just glanced by a few times and assumed the issue was handled with the information.

**Review by dberenbaum — APPROVED** [review]
LGTM. @rogermparent I approved https://github.com/iterative/gatsby-theme-iterative/pull/89. Do you want to merge that first and then this one can be merged?

**Comment by yathomasi** [comment]
The theme is updated and merged into this PR as well. The changes should be reflected on the preview deployment.

**Comment by rogermparent** [comment]
image
From the preview :tada:

**Pull request merged**